File: golang/2501.longest-square-streak-in-an-array.go
func longestSquareStreak(nums []int) int {
	const limit = int(1e5)

	cache := make(map[int]bool, len(nums))
	seen := make(map[int]bool)

	for _, num := range nums {
		cache[num] = true
	}

	maxLen := 0

	for _, num := range nums {
		if _, ok := seen[num]; ok {
			continue
		}
		seen[num] = true

		cur := num
		curLen := 1

		for cur*cur <= limit && cache[cur*cur] {
			cur *= cur
			curLen++
		}

		if curLen > maxLen {
			maxLen = curLen
		}
	}

	if maxLen < 2 {
		return -1
	}

	return maxLen
}

File: cpp/2501.longest-square-streak-in-an-array.cpp
#include <unordered_set>
#include <vector>

using std::vector, std::unordered_set;

class Solution {
public:
  int longestSquareStreak(vector<int> &nums) {
    unordered_set<int> cache(nums.begin(), nums.end());
    unordered_set<int> seen;
    const int limit = int(1e5);

    int longest = 0;

    for (int num : nums) {
      if (seen.count(num)) {
        continue;
      }

      seen.insert(num);
      long cur = num;
      int curLen = 1;

      while (cur * cur <= limit && cache.count(cur * cur)) {
        curLen++;
        cur *= cur;
      }

      if (curLen > longest) {
        longest = curLen;
      }
    }

    if (longest < 2) {
      return -1;
    }

    return longest;
  }
};